### DISCUSSION TITLE
Unclear error for an optional field in schema with a check

### DISCUSSION THREAD

**Discussion by mehemken** [kcl-lang/community#22] [Q&A]
Hello,
I'm having a difficult user experience. I've only been using kcl since this week, so it is possible user error.

I've used kcl-openapi to generate the model for fluxcd HelmRelease v2beta2. The relevant configuration looks like this

```
schema HelmToolkitFluxcdIoV2beta2HelmReleaseSpec:
    """
    HelmReleaseSpec defines the desired state of a Helm release.

    Attributes
    ----------
    releaseName : str, default is Undefined, optional
        ReleaseName used for the Helm release. Defaults to a composition of '[TargetNamespace-]Name'.
    """

    releaseName?: str

    check:
        len(releaseName) <= 53
        len(releaseName) >= 1
```

My use case is that I'm creating an instance of HelmRelease with some default values that all of my HelmReleases will use. Then I am creating specific HelmReleases that inherit from my "default" HelmRelease.

The trouble happens when I don't specify optional fields that have checks. This is just an example with `releaseName`. It is not a required field, but kcl throws the error below. If I set that field in my "default" HelmRelease the error goes away and kcl shows a similar error for other optional fields that I haven't defined.

As they are optional fields, how do I avoid this error without providing a value?

```
Error: failed to compile the kcl package
EvaluationError
   --> /.../local-kind/flux/models/helm/release/v2beta2/helm_toolkit_fluxcd_io_v2beta2_helm_rel
ease.k:138:1
    |
138 |         len(releaseName) <= 53
    |  object of type 'UndefinedType' has no len()
```

**Comment by mehemken** [comment]
If I remove all the `len(fieldName)` checks, I get rid of the errors. Again, I've only been using kcl since this week so I might not understand something about kcl.

Is this the intended behavior, or is it a bug?

**Reply by RickardCardell** [reply]
You can have a guard:
```
    check:
        len(releaseName) <= 53 if releaseName != Undefined
        len(releaseName) >= 1 if releaseName != Undefined
```
Then that check is only activated if `releaseName` is set

**Comment by WaldoBlackwood** [comment]
> Hello, I'm having a difficult user experience. I've only been using kcl since this week, so it is possible user error.
> 
> I've used kcl-openapi to generate the model for fluxcd HelmRelease v2beta2. The relevant configuration looks like this
> 
> ```
> schema HelmToolkitFluxcdIoV2beta2HelmReleaseSpec:
>     """
>     HelmReleaseSpec defines the desired state of a Helm release.
> 
>     Attributes
>     ----------
>     releaseName : str, default is Undefined, optional
>         ReleaseName used for the Helm release. Defaults to a composition of '[TargetNamespace-]Name'.
>     """
> 
>     releaseName?: str
> 
>     check:
>         len(releaseName) <= 53
>         len(releaseName) >= 1
> ```
> 
> My use case is that I'm creating an instance of HelmRelease with some default values that all of my HelmReleases will use. Then I am creating specific HelmReleases that inherit from my "default" HelmRelease.
> 
> The trouble happens when I don't specify optional fields that have checks. This is just an example with `releaseName`. It is not a required field, but kcl throws the error below. If I set that field in my "default" HelmRelease [Rabbit Road game](https://rabbitroads.uk/) the error goes away and kcl shows a similar error for other optional fields that I haven't defined.
> 
> As they are optional fields, how do I avoid this error without providing a value?
> 
> ```
> Error: failed to compile the kcl package
> EvaluationError
>    --> /.../local-kind/flux/models/helm/release/v2beta2/helm_toolkit_fluxcd_io_v2beta2_helm_rel
> ease.k:138:1
>     |
> 138 |         len(releaseName) <= 53
>     |  object of type 'UndefinedType' has no len()
> ```

This happens because the check block is still evaluated even when the field is undefined. Since releaseName is optional, its value is Undefined, and calling len on Undefined causes the error. KCL does not automatically skip checks for optional fields.

You need to guard the check so it only runs when the value exists. In other words, test for defined first. For example, wrap your checks in a condition like “if releaseName is not Undefined”. That way, the length rules only apply when the user actually sets the field.

Optional fields with checks must always be written defensively. Otherwise, KCL assumes the value is real and tries to evaluate it. This is expected behavior, but the error message is not very friendly for new users.
